portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java

```diff
@@ -86,16 +100,63 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
 
     /* Get the input from the form. */
+    String name = request.getParameter("name-input");
     String comment = request.getParameter("text-input");
+    String imageUrl = getUploadedFileUrl(request, "image");
+
     if (comment != null) {
       Entity commentEntity = new Entity("Comment");
-      comment.replaceAll("(?i)<(/?script[^>]*)>", "&lt;$1&gt;");
-      commentEntity.setProperty("comment", comment);
+      commentEntity.setProperty("name", safeServerEncoding(name));
+      commentEntity.setProperty("comment", safeServerEncoding(comment));
+      commentEntity.setProperty("imageUrl", imageUrl);
 
       DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
       datastore.put(commentEntity);
     }
     
     response.sendRedirect("/index.html");
   }
+
+  /* Replace script tags. */
+  private String safeServerEncoding(String text) {
+    return text.replaceAll("(?i)<(/?script[^>]*)>", "&lt;$1&gt;");
+  }
+
+  /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file. */
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+
+    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    // We could check the validity of the file here, e.g. to make sure it's an image file
+    // https://stackoverflow.com/q/10779564/873165
+
+    // Use ImagesService to get a URL that points to the uploaded file.
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
+    ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+
+    // To support running in Google Cloud Shell with AppEngine's dev server, we must use the relative
+    // path to the image, rather than the path returned by imagesService which contains a host.
+    try {
+      URL url = new URL(imagesService.getServingUrl(options));
+      return url.getPath();
+    } catch (MalformedURLException e) {
+      return imagesService.getServingUrl(options);
```

Comment:
> Better: Distinguish between dev and prod environment in code:

I second that -- I assume the returned URL is not valid in case a MalformedURLException occurs in the prod environment?